src/xcode/ENA/ENA/Source/Services/Risk/Provider/RiskProvider.swift

```diff
@@ -107,11 +108,28 @@ final class RiskProvider: RiskProviding {
 
 			// Keep downloading key packages and trace warning packages for plausible deniability
 
-			downloadKeyPackages()
-
-			appConfigurationProvider.appConfiguration().sink { [weak self] appConfiguration in
-				self?.downloadTraceWarningPackages(with: appConfiguration, completion: { _ in })
-			}.store(in: &subscriptions)
+			downloadKeyPackages { [weak self] _ in
+				guard let self = self else {
+					return
+				}
+				self.appConfigurationProvider.appConfiguration().sink { [weak self] appConfiguration in
+					self?.downloadTraceWarningPackages(with: appConfiguration, completion: { [weak self] result in
+						guard let self = self else {
+							return
+						}
+						switch result {
+						case .success:
+							if let risk = self.previousRiskIfExistingAndNotExpired(userInitiated: userInitiated) {
+								self.successOnTargetQueue(risk: risk)
+							} else {
+								assertionFailure("This should never happen. At this point there should be a risk to use available.")
+							}
+						case .failure(let error):
+							self.failOnTargetQueue(error: error)
```

Comment:
Why does a failed trace warning package call fail, but not a failed key package download?
In general I'm not sure if its a good pattern, to have several completion calls (maybe even failure and success) for 1 risk request.

Reply:
I missed that its in a guard with return. Cleared.